lib/si-cli/src/engine/podman_engine.rs

```diff
@@ -1,18 +1,62 @@
 use crate::engine::{ContainerEngine, ContainerReleaseInfo, SiContainerSummary, SiImageSummary};
-use crate::{CliResult, SiCliError};
+use crate::{CliResult, SiCliError, CONTAINER_NAMES};
 use async_trait::async_trait;
+use color_eyre::eyre::eyre;
+use futures::StreamExt;
+use indicatif::{MultiProgress, ProgressBar, ProgressStyle};
+use podman_api::models::{ContainerMount, Namespace, PerNetworkOptions, PortMapping};
+use podman_api::opts::{
+    ContainerCreateOpts, ContainerDeleteOpts, ContainerListFilter, ContainerListOpts,
+    ContainerLogsOpts, ContainerStopOpts, ImageListOpts, PullOpts,
+};
 use podman_api::Podman;
+use std::collections::HashMap;
 use std::path::PathBuf;
 
 pub struct PodmanEngine {
     podman: Podman,
+    network: String,
 }
 
 impl PodmanEngine {
     #[allow(clippy::new_ret_no_self)]
-    pub async fn new(_sock: Option<String>) -> CliResult<Box<dyn ContainerEngine>> {
-        let podman = Podman::unix("//run/podman.sock");
-        Ok(Box::new(PodmanEngine { podman }))
+    pub async fn new(sock: Option<String>) -> CliResult<Box<dyn ContainerEngine>> {
+        let podman_sock = if let Some(sock) = sock {
+            sock
+        } else {
+            "".to_string()
+        };
+
+        let podman_socket_candidates = vec![
+            #[allow(clippy::disallowed_methods)]
+            // To determine a path to the user's runtime directory.
+            std::path::Path::new(&std::env::var("XDG_RUNTIME_DIR")?)
+                .join("podman")
+                .join("podman.sock"),
+            std::path::Path::new("/var/run/podman.sock").to_path_buf(),
+        ];
+
+        let podman: Podman;
+        if let "" = podman_sock.as_str() {
+            let socket = podman_socket_candidates
+                .iter()
+                .find(|candidate| candidate.exists())
+                .ok_or(eyre!(
+            "failed to determine podman socket location. Set a custom location using `--podman-sock` \
+            or `SI_PODMAN_SOCK`; candidates={podman_socket_candidates:?}"
+        ))?;
+            podman = Podman::unix(socket)
+        } else {
+            println!("Checking for user supplied podman.sock");
+            let path = std::path::Path::new(podman_sock.as_str()).to_path_buf();
+            podman = Podman::unix(path);
+        }
+
+        Ok(Box::new(PodmanEngine {
+            podman,
+            // TODO: Create this network automatically.
+            network: "si".to_owned(),
```

Comment:
I want to get this implemented before merging. Right now there isn't a great place to perform this network creation and deletion logic. I was thinking about updating the `ContainerEngine` trait to take two new methods (e.g., `pre_hook`, `post_hook`) and placing the network creation and deletion logic in there.

What do you all think?

Reply:
I added `create_network` and `delete_network` methods to the `ContainerEngine` trait to support this network creation and deletion logic.